src/Shared/Buffers.MemoryPool/SlabMemoryPool.cs

```diff
@@ -160,6 +160,21 @@ internal void Return(MemoryPoolBlock block)
             }
         }
 
+        // This method can ONLY be called from the finalizer of MemoryPoolBlock
+        internal void RefreshBlock(MemoryPoolSlab slab, int offset, int length)
+        {
+            lock (_disposeSync)
```

Comment:
What happens if the GC suspended all other threads while the MemoryPoolBlock finalizer runs and one of those suspended threads has the _disposeSync lock? Normally the _disposeSync lock isn't taken for long, but if the thread was suspended, I could see it potentially deadlocking.

@Maoni0 Can you safely acquire a lock in a finalizer?

Reply:
I found https://blogs.msdn.microsoft.com/yunjin/2005/07/05/special-threads-in-clr/

>Because there is only one thread to run all finalizers, if one finalizer is blocked, no other finalizers could run. So it is discouraged to take any lock in finalizer. Also see [Maoni Stephens's blog](http://blogs.msdn.com/maoni/archive/2004/11/04/252697.aspx) for details about finalizer thread.

It recommends not taking any lock, but only because it can stop other finalizers from running. It doesn't say whether the GC suspends normal threads while the finalizer is running.

[This doc](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals#concurrent-garbage-collection) seems to indicate the GC will not suspend other threads .NET Core unless concurrent garbage collection is disabled. That said, as a server, we need to make sure we don't deadlock when concurrent garbage connection is disabled considering we recommend disabling it "[t]o improve performance when several processes are running, disable concurrent garbage collection."

@Maoni0 Please feel free to correct any inaccuracies in my understanding.

Reply:
Further reading now leads me to believe that finalizers don't run "during" a GC where the runtime might suspend threads, but that it's merely triggered by the collection adding objects-waiting-to-be-finalized to the finalizer queue. So the object being finalized is still alive as far as GC is concerned until the object is dequeued from the finalizer queue by the finalizer thread and the object is no longer referenced (that is *if* the object is no longer referenced). When collections are actually run, even the finalizer thread is suspended at least if concurrent garbage collection is disabled.

This is making me think that the effort we go through to reconstruct the MemoryPoolBlock might not be worth it, and that we should just add `this` back to the pool in the finalizer and "resurrect" the MemoryPoolBlock.

I'm not writing this long explanation, because I'm confident I'm correct. I just want to explain my current understanding so the experts can clear up what I'm wrong about.

Reply:
sorry I had not gotten around to responding to this - to GC the finalizer thread is no different from any other managed threads - it will be suspended when GC calls SuspendEE to suspend managed threads.

Reply:
Thanks!